### PR TITLE
Unconditionally leak autograd threads.

### LINKED ISSUE #48888: Background threads not exiting after calling loss.backward() on Win7
## 🐛 Bug

Background threads not exiting after calling loss.backward() on Win7, forced to kill the python console between each run.

## To Reproduce

1. Run conda with python 3.8 on windows 7
2. Run on either cpu or gpu (both tried)
3. Install from wheel, pip or build from source (all three tried)
3. Calculate a loss based on your loss function
4. Call loss.backward()
5. Program will continue until the end of the script but never give back the command prompt

```
import torch
import torch.nn as nn
import numpy as np

dtype = torch.float
device = torch.device("cpu")

class MLP(nn.Module):
    def __init__(self, inshape, outshape):
        super(MLP, self).__init__()
        self.layers = nn.Sequential(
            nn.Linear(inshape, outshape)
        )

    def forward(self, x):
        x = x.view(x.size(0), -1)
        x = self.layers(x)
        return x

data = torch.from_numpy(np.random.rand(1, 10).astype(np.float32))
target = torch.from_numpy(np.random.rand(1, 1).astype(np.float32))

model = MLP(10, 1)
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
loss_fn = nn.MSELoss()

predict = model(data)
loss = loss_fn(predict, target)

model.zero_grad()
loss.backward()
optimizer.step()

print("Program done, command prompt expected below.")
```
returns
```
(PY38) D:\GitData>C:/Users/Grinay/Anaconda3/envs/PY38/python.exe d:/GitData/bitnet/train_net.py
Program done, command prompt expected below.

```
and nothing happens until I kill the python terminal.

## Expected behavior

Expected behaviour is that the code finishes to run until done and then returns to the command prompt.

## Environment

```
(PY38) D:\GitData\pytorch\torch\utils>python collect_env.py
Collecting environment information...
PyTorch version: 1.8.0a0+6317e0b
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 7 Professional N
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.18.2

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 780
Nvidia driver version: 441.22
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\bin\cudnn64_7.dll
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] torch==1.8.0a0
[pip3] torchaudio==0.7.0
[pip3] torchvision==0.8.1
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              h74a9793_1
[conda] mkl                       2019.0                   pypi_0    pypi
[conda] mkl-include               2020.2                      256
[conda] mkl-service               2.3.0            py38h2bbff1b_0
[conda] mkl_fft                   1.2.0            py38h45dec08_0
[conda] mkl_random                1.1.1            py38h47e9c7a_0
[conda] numpy                     1.19.4                   pypi_0    pypi
[conda] numpy-base                1.19.2           py38ha3acd2a_0
[conda] torch                     1.8.0a0                  pypi_0    pypi
[conda] torchaudio                0.7.0                      py38    pytorch
[conda] torchvision               0.8.1                py38_cu102    pytorch

(PY38) D:\GitData\pytorch\torch\utils>
```

 - PyTorch Version (e.g., 1.0): Multiple versions tested
 - OS (e.g., Linux): Win7 Pro N
 - How you installed PyTorch (`conda`, `pip`, source): conda, pip wheel and source tested
 - Build command you used (if compiling from source): Ninja with VS2019, cuda, cudnn (Edit: and compute_arch 3.5 for my GPU)
 - Python version: 3.8 (I think I tested with 3.7 as well, I do not remember though)
 - CUDA/cuDNN version: 10.2/7.6.XXX, (Edit: also tested from wheel and conda without cuda/cudnn)
 - GPU models and configuration: GTX780 with the drivers shipped with CUDA 10.2 (Edit: also tested with the newest graphic drivers)
 - Any other relevant information: Sorry, none.

## Additional context

Edit: Commenting out the loss.backward() call in the code-snippet will make the program exit normally.

cc @peterjc123 @maxluk @nbcsm @guyang3532 @gunandrose4u @mszhanyi @skyline75489

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53018 Unconditionally leak autograd threads.**

Fixes #48888, maybe.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26725185](https://our.internmc.facebook.com/intern/diff/D26725185)

### CODE REVIEW

**Comment by albanD** [comment]
Do we have somewhere the reason why this logic was added in the first place?
This question was never answered on the PR that adds it https://github.com/pytorch/pytorch/pull/21438#discussion_r299044331 and there is no comment there explaining why we need this.

**Comment by ezyang** [comment]
@albanD It's a hygiene thing. In an ideal world, you properly cleanup everything after yourself on shutdown. This way, if you e.g., load PyTorch dynamic library and then unload it, you end up with a net zero state; if you're leaking threads, this isn't true.

As you might imagine, this is a pretty niche use case. So since it's causing problems it's more trouble than it's worth. Looks like need to disable the relevant tests.

**Review by Balandat — APPROVED** [review]
This indeed fixes an issue for me with test binaries not properly shutting down after a successful test run. I can't really speak to the larger blast radius of this change though.

**Comment by malfet** [comment]
@albanD I remember I had to fix logic in the shutdown sequence because it lead to occasional crashes (especially while running C++ tests)
Perhaps a reasonable solution would be to introduce an environment variable controllable timeout during shutdown (setting it say 10 sec sounds reasonable by default)

**Comment by ezyang** [comment]
@malfet Whatever logic that was, this PR is bypassing it entirely because we no longer send the shutdown tasks. Are you suggesting we should avoid landing this task (so that we keep exercising the shutdown logic), or that someone should look into doing this later?

**Comment by malfet** [comment]
@ezyang Can you please hold of landing this one and I will send PR with the timeout logic I have in mind ( that will still leak threads, but after the delay)

**Comment by malfet** [comment]
How about something along the following lines: https://github.com/pytorch/pytorch/pull/53170
(With that PR invoking code with TORCH_AUTOGRAD_SHUTDOWN_WAIT_LIMIT=0 would be equivalent to current PR, but by default shutdown would be performed correctly)
